opm/simulators/linalg/DILU.hpp

```diff
@@ -0,0 +1,210 @@
+ */
+template <class M, class X, class Y>
+class SeqDilu : public PreconditionerWithUpdate<X, Y>
+{
+public:
+    //! \brief The matrix type the preconditioner is for.
+    typedef M matrix_type;
+    //! \brief The domain type of the preconditioner.
+    typedef X domain_type;
+    //! \brief The range type of the preconditioner.
+    typedef Y range_type;
+    //! \brief The field type of the preconditioner.
+    typedef typename X::field_type field_type;
+    //! \brief scalar type underlying the field_type
+#if DUNE_VERSION_NEWER(DUNE_ISTL, 2, 7)
+    typedef Simd::Scalar<field_type> scalar_field_type;
+#else
+    typedef SimdScalar<field_type> scalar_field_type;
+#endif
+
+    /*! \brief Constructor.
+       Constructor gets all parameters to operate the prec.
+       \param A The matrix to operate on.
+     */
+    SeqDilu(const M& A)
+        : A_(A)
+    {
+        CheckIfDiagonalPresent<M, 1>::check(A_);
+        // we build the inverseD matrix
+        Dinv_.resize(A_.N());
+
+        // is this the correct usage?
+        update();
+    }
+
+    virtual void update() override {
+
+        using block = typename M::block_type;
+        
+        for ( auto i = A_.begin(); i != A_.end(); ++i)
+        {
+            block Dinv_temp;
+            for (auto a_ij=i->begin(); a_ij != i->end(); ++a_ij) {
+                auto a_ji = A_[a_ij.index()].find(i.index());
+                
+                if (a_ij.index() == i.index()) {
+                    Dinv_temp += A_[i.index()][i.index()];
+                }
+
+                // if A[i, j] != 0 and A[j, i] != 0:
+                else if (a_ji != A_[a_ij.index()].end() && (a_ij.index() < i.index())) {
+
+                    auto d_i = Dinv_[a_ij.index()];
+                    d_i.rightmultiply(*a_ji);
+                    d_i.leftmultiply(*a_ij);
+
+                    // d[j] -= A[j, i] * d[i] * A[i, j]
+                    Dinv_temp -= d_i;
+                }
+            }
+            Dinv_temp.invert();
+            Dinv_[i.index()] = Dinv_temp;
+        }
+    }
+
+    /*!
+       \brief Prepare the preconditioner.
+       \copydoc Preconditioner::pre(X&,Y&)
+     */
+    virtual void pre(X& v, Y& d) override
+    {
+        DUNE_UNUSED_PARAMETER(v);
+        DUNE_UNUSED_PARAMETER(d);
+    }
+
+
+  /*!
+       \brief Apply the preconditioner.
+       \copydoc Preconditioner::apply(X&,const Y&)
+     */
+ virtual void apply(X& v, const Y& d) override
+    {
+        
+    // M = (D + L_A) D^-1 (D + U_A)   (a LU decomposition of M)
+    // where L_A and U_A are the strictly lower and upper parts of A and M has the properties:
+    // diag(A) = diag(M)
+    // solving the product M^-1(b-Ax) using upper and lower triangular solve
+    // z = x_k+1 - x_k = M^-1(b - Ax) = (D + L_A)^-1 D (D + U_A)^-1 (b - Ax)
+
+    typedef typename Y::block_type dblock;
+    
+    // copy current v
+    X x(v);
+    X z;
+    Y y;
+    z.resize(A_.N());
+    y.resize(A_.N());
```

Comment:
These three vectors have a known size and should only be allocated once, probably in the constructor.

Reply:
Yes you are right. This was initially used to ease the implementation but in the optimised version I removed the need for storing temporarily vectors completely, since we are only calculating the update to x.